core/src/main/java/com/yahoo/imapnio/async/data/StoreResult.java

```diff
@@ -0,0 +1,99 @@
+package com.yahoo.imapnio.async.data;
+
+import java.util.List;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+import com.sun.mail.imap.protocol.IMAPResponse;
+
+/**
+ * This class provides the highest modification sequence, the list of IMAP responses, and the list of modified messages number
+ * from store command response based on RFC3501 and RFC7162 as following.
+ *
+ * <pre>
+ * resp-text-code      =/ "HIGHESTMODSEQ" SP mod-sequence-value /
+ *                        "MODIFIED" SP sequence-set
+ * </pre>
+ */
+public class StoreResult {
+
+    /** The highest modification sequence. */
+    private Long highestModSeq;
+
+    /** The collection of IMAP responses. */
+    private final List<IMAPResponse> imapResponses;
+
+    /** The collection of message number as sequence-set, only shown when CondStore is enabled. */
+    private final MessageNumberSet[] modifiedMsgSets;
+
+    /**
+     * Initializes a {@code StoreResult} object with IMAP responses collection.
+     *
+     * @param imapResponses collection of IMAP responses from store command result
+     */
+    public StoreResult(@Nonnull final List<IMAPResponse> imapResponses) {
+        this.highestModSeq = null;
+        this.imapResponses = imapResponses;
+        this.modifiedMsgSets = null;
+    }
+
+    /**
+     * Initializes a {@code StoreResult} object with IMAP responses collection and modified message number collection.
+     *
+     * @param imapResponses collection of IMAP responses from store command result
+     * @param modifiedMsgSets collection of modified message number from store command result
+     */
+    public StoreResult(@Nonnull final List<IMAPResponse> imapResponses, @Nonnull final MessageNumberSet[] modifiedMsgSets) {
+        this(null, imapResponses, modifiedMsgSets);
+    }
+
+    /**
+     * Initializes a {@code StoreResult} object with the highest modification sequence and IMAP responses collection.
+     *
+     * @param highestModSeq the highest modification from store command result
+     * @param imapResponses collection of IMAP responses from store command result
+     */
+    public StoreResult(@Nonnull final Long highestModSeq, @Nonnull final List<IMAPResponse> imapResponses) {
+        this(highestModSeq, imapResponses, null);
+    }
+
+    /**
+     * Initializes a {@code StoreResult} object with the highest modification sequence, IMAP responses collection,
+     * and modified message number collection.
+     *
+     * @param highestModSeq the highest modification from store command result
+     * @param imapResponses collection of IMAP responses from store command result
+     * @param modifiedMsgSets collection of modified message number from store command result
+     */
+    public StoreResult(@Nullable final Long highestModSeq, @Nonnull final List<IMAPResponse> imapResponses,
```

Comment:
One general idea for annotation is
(1) If you choose to have one constructor for all with null/nonnull, then you can have nullable here
(2) Otherwise, if you choose to have multiple constructor, each one's constructor variable should be nonnull. It is because in this case, people cant pass in null. If they want pass in null, they should use other constructors.